src/Platform/Microsoft.Testing.Extensions.AzureDevOps/AzureDevOpsReporter.cs

```diff
@@ -0,0 +1,177 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Testing.Extensions.Reporting;
+using Microsoft.Testing.Extensions.Reporting.Resources;
+using Microsoft.Testing.Platform;
+using Microsoft.Testing.Platform.CommandLine;
+using Microsoft.Testing.Platform.Extensions.Messages;
+using Microsoft.Testing.Platform.Extensions.OutputDevice;
+using Microsoft.Testing.Platform.Extensions.TestHost;
+using Microsoft.Testing.Platform.Helpers;
+using Microsoft.Testing.Platform.OutputDevice;
+
+namespace Microsoft.Testing.Extensions.TrxReport.Abstractions;
+
+internal sealed class AzureDevOpsReporter :
+    IDataConsumer,
+    IDataProducer,
+    IOutputDeviceDataProducer
+{
+    private readonly IOutputDevice _outputDisplay;
+
+    private static readonly char[] NewlineCharacters = new char[] { '\r', '\n' };
+    private readonly ICommandLineOptions _commandLine;
+    private readonly IEnvironment _environment;
+    private string _severity = "error";
+
+    public AzureDevOpsReporter(
+        ICommandLineOptions commandLine,
+        IEnvironment environment,
+        IOutputDevice outputDisplay)
+    {
+        _commandLine = commandLine;
+        _environment = environment;
+        _outputDisplay = outputDisplay;
+    }
+
+    public Type[] DataTypesConsumed { get; } =
+    [
+        typeof(TestNodeUpdateMessage)
+    ];
+
+    public Type[] DataTypesProduced { get; } = [typeof(SessionFileArtifact)];
+
+    /// <inheritdoc />
+    public string Uid { get; } = nameof(AzureDevOpsReporter);
+
+    /// <inheritdoc />
+    public string Version { get; } = AppVersion.DefaultSemVer;
+
+    /// <inheritdoc />
+    public string DisplayName { get; } = AzureDevOpsResources.DisplayName;
+
+    /// <inheritdoc />
+    public string Description { get; } = AzureDevOpsResources.Description;
+
+    /// <inheritdoc />
+    public Task<bool> IsEnabledAsync()
+    {
+        bool isEnabled = _commandLine.IsOptionSet(AzureDevOpsCommandLineOptions.AzureDevOpsOptionName)
+            && string.Equals(_environment.GetEnvironmentVariable("TF_BUILD"), "true", StringComparison.OrdinalIgnoreCase);
+
+        if (isEnabled)
+        {
+            bool found = _commandLine.TryGetOptionArgumentList(AzureDevOpsCommandLineOptions.AzureDevOpsReportSeverity, out string[]? arguments);
+            if (found && arguments?.Length > 0)
+            {
+                if (string.Equals(arguments[0], "warning", StringComparison.OrdinalIgnoreCase))
+                {
+                    _severity = "warning";
+                }
+            }
+        }
+
+        return Task.FromResult(isEnabled);
+    }
+
+    public async Task ConsumeAsync(IDataProducer dataProducer, IData value, CancellationToken cancellationToken)
+    {
+        if (cancellationToken.IsCancellationRequested)
+        {
+            return;
```

Comment:
While this is a pattern we are following in other places, I don't feel like it's correct. I think we should call `ThrowIfCancellationIsRequested`. But whether or not the cancellation exception is handled correctly for now is something I don't really know.

Reply:
good point, I think you made a separate issue for this, let's solve it there. I don't want to introduce unrelated changes here.